test/utils/CallUtils.sol

```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+import {Call} from "../../src/libraries/CallLib.sol";
+import {IKeyManagement} from "../../src/interfaces/IKeyManagement.sol";
+import {IERC7821} from "../../src/interfaces/IERC7821.sol";
+import {Settings} from "../../src/libraries/SettingsLib.sol";
+import {TestKeyManager, TestKey} from "./TestKeyManager.sol";
+
+/// @dev A wrapper around Call that includes callback data for processing after execution
+struct HandlerCall {
+    Call call;
+    bytes callback;
+}
+
+/// @dev Utility library for Call and HandlerCall objects
+library CallUtils {
+    using CallUtils for Call;
+    using CallUtils for Call[];
+    using CallUtils for HandlerCall;
+    using CallUtils for HandlerCall[];
+    using TestKeyManager for TestKey;
+
+    // Constants
+    address constant SELF_CALL = address(0);
+    bytes32 internal constant BATCHED_CALL = 0x0100000000000000000000000000000000000000000000000000000000000000;
+
+    // Call array operations
+
+    /// @dev Create empty Call array
+    function init() internal pure returns (Call[] memory) {
+        return new Call[](0);
+    }
+
+    /// @dev Add a call to an array
+    function push(Call[] memory calls, Call memory call) internal pure returns (Call[] memory) {
+        Call[] memory newCalls = new Call[](calls.length + 1);
+        for (uint256 i = 0; i < calls.length; i++) {
+            newCalls[i] = calls[i];
+        }
+        newCalls[calls.length] = call;
+        return newCalls;
+    }
+
+    // Call manipulation
+
+    /// @dev Create default empty Call
+    function initDefault() internal pure returns (Call memory) {
+        return Call({to: address(0), value: 0, data: ""});
+    }
+
+    function withTo(Call memory call, address to) internal pure returns (Call memory) {
+        call.to = to;
+        return call;
+    }
+
+    function withValue(Call memory call, uint256 value) internal pure returns (Call memory) {
+        call.value = value;
+        return call;
+    }
+
+    function withData(Call memory call, bytes memory data) internal pure returns (Call memory) {
+        call.data = data;
+        return call;
+    }
+
+    // Call encoding for specific operations
+
+    /// @dev Add execute call to array
+    function addExecute(Call[] memory calls, Call[] memory innerCalls) internal pure returns (Call[] memory) {
+        return push(calls, encodeExecuteCall(innerCalls));
+    }
+
+    /// @dev Create call to execute a batch
+    function encodeExecuteCall(Call[] memory calls) internal pure returns (Call memory) {
```

Comment:
it would be nice if this could handle different modes i think

Reply:
Yea its tricky because other modes require opData, which is encoded differently